api/envoy/config/bootstrap/v3/bootstrap.proto

```diff
@@ -181,6 +181,15 @@ message Bootstrap {
   // Optional set of stats sinks.
   repeated metrics.v3.StatsSink stats_sinks = 6;
 
+  // To save memory and CPU consumption on blocks of stats that are never referenced throughout
+  // the process lifetime, they can be encapsulated in a DeferredCreationCompatibleInterface. Then
+  // the Envoy bootstrap configuration can be set to defer the instantiation of those blocks. Note that
+  // when the blocks of stats are created, they carry an extra overhead (around 60-100 bytes depending
+  // on worker thread count) due to internal bookkeeping data structures.
+  //
+  // When set to true, Envoy will defer initializing selected blocks of stats until the first time they are updated.
+  bool enable_deferred_creation_stats = 39;
```

Comment:
Will we ever want to free the deferred-creation stats if they aren't used for some period of time? If so, would it change how we structure this config? Or would there be any other related config for this setting?

Would it be worth having an empty message here, which when set enables this feature? Then we could put additional config for it in that message if/when we need it, such as options for selecting which stats are deferred-creation, timer for freeing them after non-use, etc.

Reply:
@jmarantz as well. 

My thinking is that this is a all-or-none setting that controls the deferred creation characteristics for all stats that's supports deferred creation, but I'd let Josh to chime in here.

Reply:
I like Greg's suggestion: by making this a message with a bool in it, rather than just a bool, you present an API that can be extended in the future to maybe include a timeout.

I think a timeout would be useful in situations like the one you have in mind, but where a cluster might get a request, bringing in all its stats, and then not get any for many days after that. We might prefer to drop those stats then have them consume memory until process exit.

Reply:
Sounds good, will change it to a submessage field. 
Is there any preference on how to name the new sub-message? 

one QQ: is the timeout idea also applicable to the "current " non-deferred-created stats, that are always populated, but not used?

Reply:
Maybe something like `SparseStatAllocation` for the name?

I don't think the timeout is applicable to current stats, unless we wanted to change the definition of `used()` to be `used within last X duration`.

Reply:
Thanks ALL!

@jmarantz proposed to use DeferredStatOptions in the bootstrap config. 
PR updated, PTAL.

Reply:
I have a slight preference for DeferredStatOptions but I'm OK with SparseStatAllocation as well. I thought we had gone with Deferred in the code in lieu of Lazy, and it would be preferable to be consistent. But I don't feel too strongly about it.

Reply:
FYI: format failure on protobuf; I think the placement of the nested message.

https://dev.azure.com/cncf/envoy/_build/results?buildId=140116&view=logs&j=c95d62a5-6bb9-58f0-2a02-7575fd482a15&t=76c9b3e6-0d40-57e9-06f3-2219e9228a0a